### PR TITLE
dmSDK: Added possibility to load collection proxies and spawn factories from C++

### LINKED ISSUE #8930: Spawn factories (stars)

### PR DESCRIPTION
**Pull request by JCash**
This update adds more functionality to our C++ sdk, which is part of our effort to allow a developer to write the game logic using C++.
It is an ongoing task, and we'll add more functionality along the way.

# Technical changes:

## dmsdk/extension/extension.h

Allows for more control of from within an extension:
* dmExtension::EVENT_ID_ENGINE_INITIALIZED - called at the very end of dmEngine::Init()
* dmExtension::EVENT_ID_ENGINE_DELETE - called at the very beginning of dmEngine::Delete()

## dmsdk/gameobject/gameobject_props.h

Removed Lua dependency when creating game object properties.
Added property container type to sdk for making it possible to spawn game objects:

* dmGameObject::HPropertyContainer - Used when spawning game objects/factories
* dmGameObject::HPropertyContainerBuilder - Used when creating a HPropertyContainer

## dmsdk/gameobject/gameobject.h
Added more functionality that allows for spawning game objects, and getting components.

## dmsdk/gameobject/component.h

Added more type safety (also for readability) using two new typedefs:

* dmGameObject::HComponentWorld
* dmGameObject::HComponentInternal

## Misc

Removed deprecated function:
* dmGameObject::GetComponentFromLua(), in favor of dmScript::GetComponentFromLua()

Fixes: https://github.com/defold/defold/issues/8930

## PR checklist

* [ ] Code
	* [ ] Add engine and/or editor unit tests.
	* [ ] New and changed code follows the overall code style of existing code
	* [ ] Add comments where needed
* [ ] Documentation
	* [ ] Make sure that API documentation is updated in code comments
	* [ ] Make sure that manuals are updated (in github.com/defold/doc)
* [ ] Prepare pull request and affected issue for automatic release notes generator
	* [ ] Pull request - Write a message that explains what this pull request does. What was the problem? How was it solved? What are the changes to APIs or the new APIs introduced? This message will be used in the generated release notes. Make sure it is well written and understandable for a user of Defold.
	* [ ] Pull request - Write a pull request title that in a sentence summarises what the pull request does. Do not include "Issue-1234 ..." in the title. This text will be used in the generated release notes.
	* [ ] Pull request - Link the pull request to the issue(s) it is closing. Use on of the [approved closing keywords](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue).
	* [ ] Affected issue - Assign the issue to a project. Do not assign the pull request to a project if there is an issue which the pull request closes.
	* [ ] Affected issue - Assign the "breaking change" label to the issue if introducing a breaking change.
	* [ ] Affected issue - Assign the "skip release notes" is the issue should not be included in the generated release notes.

----------

Example of a well written PR description:

1. Start with the user facing changes. This will end up in the release notes.
1. Add one of the GitHub approved closing keywords
1. Optionally also add the technical changes made. This is information that might help the reviewer. It will not show up in the release notes. Technical changes are identified by a line starting with one of these:
   1. `### Technical changes` 
   1. `Technical changes:`
   2. `Technical notes:`

```
There was a anomaly in the carbon chroniton propeller, introduced in version 8.10.2. This fix will make sure to reset the phaser collector on application startup.

Fixes #1234

### Technical changes
* Pay special attention to line 23 of phaser_collector.clj as it contains some interesting optimizations
* The propeller code was not taking into account a negative phase.
```

### CODE REVIEW

File: engine/gamesys/src/dmsdk/gamesys/components/comp_collection_proxy.h [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+//
+// You may obtain a copy of the License, together with FAQs at
+// https://www.defold.com/license
+//
+// Unless required by applicable law or agreed to in writing, software distributed
+// under the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR
+// CONDITIONS OF ANY KIND, either express or implied. See the License for the
+// specific language governing permissions and limitations under the License.
+
+#ifndef DMSDK_GAMESYS_COLLECTION_PROXY_H
```
Comment (JCash) [inline engine/gamesys/src/dmsdk/gamesys/components/comp_collection_proxy.h:15]: Currently left undocumented (e.g. until we want to present it as official?)

File: engine/gamesys/src/gamesys/components/comp_collection_proxy.cpp [hunk 2]
```diff
@@ -97,18 +104,46 @@ namespace dmGameSystem
         return dmGameObject::UPDATE_RESULT_OK;
     }
 
-    void LoadComplete(CollectionProxyComponent* proxy)
+    void LoadComplete(CollectionProxyComponent* proxy, dmGameObject::Result result)
```
Comment (JCash) [inline engine/gamesys/src/gamesys/components/comp_collection_proxy.cpp:107]: Reusing the same flow for two use cases: messages vs c++ invokation

File: engine/gamesys/src/gamesys/components/comp_factory.cpp [hunk 3]
```diff
@@ -231,13 +236,27 @@ namespace dmGameSystem
...
-    static dmGameObject::HPrototype GetPrototype(dmResource::HFactory factory, FactoryComponent* component)
+    static FactoryResource* CompFactoryGetResourceInternal(HFactoryComponent component)
+    {
+        return component->m_CustomResource ? component->m_CustomResource : component->m_Resource;
+    }
+
+    FactoryResource* CompFactoryGetResource(HFactoryWorld world, HFactoryComponent component)
     {
-        FactoryResource* resource = CompFactoryGetResource(component);
+        (void)world;
```
Comment (JCash) [inline engine/gamesys/src/gamesys/components/comp_factory.cpp:253]: While not technically needed, it seems nice to have a similar api from all the components in the form "CompMyTypeDoSomething(world, component)"

File: engine/gamesys/src/gamesys/components/comp_factory.cpp [hunk 4]
```diff
@@ -445,4 +459,25 @@ namespace dmGameSystem
...
+            dmGameObject::AssignInstanceIndex(index, instance);
+        }
+        else
+        {
+            dmGameObject::ReleaseInstanceIndex(index, collection);
+        }
+        return instance;
+    }
+
+
```
Comment (JCash) [inline engine/gamesys/src/gamesys/components/comp_factory.cpp:482]: Moved from script_factory.cpp, helping us to encapsulate the functionality more.

File: engine/gamesys/src/gamesys/components/comp_label.cpp [hunk 5]
```diff
@@ -557,11 +557,11 @@ namespace dmGameSystem
         InitParametersFromDescription(component, ddf);
     }
 
-    void* CompLabelGetComponent(const dmGameObject::ComponentGetParams& params)
+    dmGameObject::HComponent CompLabelGetComponent(const dmGameObject::ComponentGetParams& params)
```
Comment (JCash) [inline engine/gamesys/src/gamesys/components/comp_label.cpp:560]: Still a void*, but easier to understand what it's supposed to return.

File: engine/gamesys/src/gamesys/scripts/script_factory.cpp [hunk 6]
```diff
@@ -235,16 +237,51 @@ namespace dmGameSystem
      * end
      * ```
      */
+    static int FactoryComp_CreateWithMessage(lua_State* L, dmGameObject::HCollection collection, dmGameObject::HInstance sender_instance, dmMessage::URL* receiver,
```
Comment (JCash) [inline engine/gamesys/src/gamesys/scripts/script_factory.cpp:240]: This is the more surprising case: When does this actually happen?
I.e. we call factory.create, but _don't_ have a game object instance set?

I left it "as-is" for now. Potential candidate for removal imho.

File: engine/gamesys/src/gamesys/test/test_gamesys.cpp [hunk 7]
```diff
@@ -123,7 +123,7 @@ static dmGameObject::HInstance Spawn(dmResource::HFactory factory, dmGameObject:
 
 static dmGameObject::HInstance Spawn(dmResource::HFactory factory, dmGameObject::HCollection collection, const char* prototype_name, dmhash_t id)
 {
-    return Spawn(factory, collection, prototype_name, id, 0, 0, Point3(0, 0, 0), Quat(0, 0, 0, 1), Vector3(1, 1, 1));
+    return Spawn(factory, collection, prototype_name, id, 0, Point3(0, 0, 0), Quat(0, 0, 0, 1), Vector3(1, 1, 1));
```
Comment (Jhonnyg) [inline engine/gamesys/src/gamesys/test/test_gamesys.cpp:126]: is this using the new spawn function introduced in this PR?

**Review by Jhonnyg — DISMISSED** [review]
Lots of stuff! But great stuff!

File: engine/gamesys/src/gamesys/test/test_gamesys.cpp (hunk 7 above)
Comment (JCash) [reply engine/gamesys/src/gamesys/test/test_gamesys.cpp:126]: This is the local test_gamesys.cpp:Spawn (see above), and it in turn uses the new dmGameObject::Spawn which takes a HProperties.

File: engine/gameobject/src/dmsdk/gameobject/gameobject_props.h [hunk 8]
```diff
@@ -0,0 +1,195 @@
...
+        uint32_t m_NumberCount;
+        uint32_t m_HashCount;
+        uint32_t m_URLStringCount;
+        uint32_t m_URLStringSize;
+        uint32_t m_URLCount;
+        uint32_t m_Vector3Count;
+        uint32_t m_Vector4Count;
+        uint32_t m_QuatCount;
+        uint32_t m_BoolCount;
+    };
```
[the remark this answers is not in the document]
Comment (JCash) [reply engine/gameobject/src/dmsdk/gameobject/gameobject_props.h:77]: I'll keep it like this for now.

**Pull request merged**